intel-sgx/insecure-time/src/lib.rs

```
@@ -0,0 +1,861 @@
+fn cpuid(leaf: u32, eax: &mut u32, ebx: &mut u32, ecx: &mut u32, edx: &mut u32) {
+    unsafe { 
+        let res = __cpuid(leaf);
+        *eax = res.eax;
+        *ebx = res.ebx;
+        *ecx = res.ecx;
+        *edx = res.edx;
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct Ticks(AtomicU64);
+
+impl PartialEq<Ticks> for Ticks {
+    fn eq(&self, other: &Ticks) -> bool {
+        let t = self.0.load(Ordering::Relaxed);
+        let o = other.0.load(Ordering::Relaxed);
+        t.eq(&o)
+    }
+}
+
+impl PartialOrd<Ticks> for Ticks {
+    fn partial_cmp(&self, other: &Ticks) -> Option<cmp::Ordering> {
+        let t = self.0.load(Ordering::Relaxed);
+        let o = other.0.load(Ordering::Relaxed);
+        t.partial_cmp(&o)
+    }
+}
+
+impl Add for Ticks {
+    type Output = Ticks;
+
+    fn add(self, other: Self) -> Self::Output {
+        Ticks(AtomicU64::new(self.0.load(Ordering::Relaxed) + other.0.load(Ordering::Relaxed)))
+    }
+}
+
+impl Display for Ticks {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), fmt::Error> {
+        write!(f, "{} ticks", self.0.load(Ordering::Relaxed))
+    }
+}
+
+impl Ticks {
+    pub fn new(t: u64) -> Self {
+        Ticks(AtomicU64::new(t))
+    }
+
+    pub fn now() -> Self {
+        Ticks(Rdtscp::read().into())
+    }
+
+    pub fn abs_diff(&self, t1: &Ticks) -> Ticks {
+        let t0 = self.0.load(Ordering::Relaxed);
+        let t1 = t1.0.load(Ordering::Relaxed);
+
+        if t1 > t0 {
+            Ticks::new(t1 - t0)
+        } else {
+            Ticks::new(t0 - t1)
+        }
+    }
+
+    fn set(&self, t: Ticks) {
+        let t = t.0.load(Ordering::Relaxed);
+        self.0.store(t, Ordering::Relaxed);
+    }
+
+    pub fn from_duration(duration: Duration, freq: &Freq) -> Self {
+        let freq = freq.as_u64();
+        let ticks_secs = duration.as_secs() * freq;
+        let ticks_nsecs = duration.subsec_nanos() as u64 * freq / NANOS_PER_SEC as u64;
+        Ticks::new(ticks_secs + ticks_nsecs)
```

I think we need to handle potential overflow here, because `freq` is type of u64.

```suggestion
        let ticks_secs = duration.as_secs().saturating_mul(freq);
        let ticks_nsecs = (duration.subsec_nanos() as u64).saturating_mul(freq) / NANOS_PER_SEC as u64;
        Ticks::new(ticks_secs.saturating_add(ticks_nsecs))
```

---

Good point, although I'm not sure that `saturating` is what we want. If the `u64::MAX` limit is ever reached (which should already be the result of a bug obviously), then `Ticks` would remain static, e.g., not ticking anymore.

---

Letting `Ticks` stop ticking would result in incorrect computations. I think we can return an error here, and fallback to doing normal usercalls again. I'll update

---

I've created a separate commit to address this comment

---

Thanks